File: app/src/main/java/com/example/my_hospital_appointments/MedicalReport.java
package com.example.my_hospital_appointments;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.Toast;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.database.core.Repo;

import java.util.ArrayList;

public class MedicalReport extends AppCompatActivity {

    EditText editTextTitle,editTextDoctorName,editTextResult,editTextConclusion;
    private ArrayList<String> myArrayList=new ArrayList<String>();
    DatabaseReference myReport;
    private String patientEmail;
    DatabaseReference Reports;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_medical_report);

        editTextTitle=(EditText) findViewById(R.id.editTextTextTitleMedical);
        editTextDoctorName=(EditText) findViewById(R.id.editTextTextDoctorNameMedical);
        editTextResult=(EditText) findViewById(R.id.editTextTextTestResultMedical);
        editTextConclusion=(EditText) findViewById(R.id.editTextTextConclusionMedical);

        Reports=FirebaseDatabase.getInstance().getReference("MedicalReport");

        myArrayList.add("Select One");
        myReport = FirebaseDatabase.getInstance().getReference().child("Patients");
        myReport.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                for(DataSnapshot snapshot : dataSnapshot.getChildren())
                {
                    myArrayList.add(snapshot.child("emailAddress").getValue().toString());
                }

            }
            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });


        ArrayAdapter adapter1 = new ArrayAdapter(this, android.R.layout.simple_spinner_item,myArrayList);
        adapter1.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);

        Spinner spinnerEmails=(Spinner) findViewById(R.id.spinnerPatientMedical);

        spinnerEmails.setAdapter(adapter1);
        spinnerEmails.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {

                String myText=spinnerEmails.getSelectedItem().toString().trim();
                if(myText=="Select One".trim())
                {
                    //Nothing

                }
                else
                {
                    Toast.makeText(MedicalReport.this,"You have Selected "+myText,Toast.LENGTH_SHORT).show();
                    patientEmail =myText;
                }
            }
            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {

            }
        });

        Button myReport=(Button) findViewById(R.id.buttonSendReportMedical);
        myReport.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(patientEmail.isEmpty())
                {
                    Toast.makeText(MedicalReport.this, "FIRST SELECT A PATIENT FROM ABOVE", Toast.LENGTH_SHORT).show();
                }
                else
                {
                    setReport(patientEmail);
                }

            }
        });
    }

    private void setReport(String myKey)
    {

        String emailKey="";
        int Counter=myKey.length();
        for(int a=0; a<Counter; a++)
        {
            if(myKey.charAt(a)=='@')
            {
                break;
            }
            else
            {
                emailKey=emailKey+myKey.charAt(a);
            }
        }
        emailKey=emailKey.trim();

        String title=editTextTitle.getText().toString().trim();
        String doctorName=editTextDoctorName.getText().toString().trim();
        String testResult=editTextResult.getText().toString().trim();
        String  conclusion=editTextConclusion.getText().toString().trim();

        if(title.isEmpty())
        {
            editTextTitle.setError("REPORT TITLE REQUIRED");
            editTextTitle.requestFocus();
        }
        else if(doctorName.isEmpty())
        {
            editTextDoctorName.setError("DOCTOR NAME REQUIRED");
            editTextDoctorName.requestFocus();
        }
        else if(testResult.isEmpty())
        {
            editTextResult.setError("TEST RESULT CANNOT BE BLANK");
            editTextResult.requestFocus();
        }
        else if(conclusion.isEmpty())
        {
            editTextConclusion.setError("CONCLUSION IS REQUIRED");
            editTextConclusion.requestFocus();
        }
        else
        {
            String finalEmailKey = emailKey;
            Reports.child(emailKey).get().addOnCompleteListener(new OnCompleteListener<DataSnapshot>() {
                @Override
                public void onComplete(@NonNull Task<DataSnapshot> task) {
                    if(task.isSuccessful())
                    {
                        //String title, String doctorName, String testResult, String conclusion
                        ReportClass myReportsClass=new ReportClass(title,doctorName,testResult,conclusion);
                        Reports.child(finalEmailKey).setValue(myReportsClass);
                        //Toast.makeText(, "", Toast.LENGTH_SHORT).show();
                        Toast.makeText(MedicalReport.this, "REPORT SAVED SUCCESSFULLY", Toast.LENGTH_SHORT).show();
                    }
                }
            });
        }
    }
}

File: app/src/main/java/com/example/my_hospital_appointments/MedicalReport.kt
package com.example.my_hospital_appointments

import android.os.Bundle
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import com.google.android.gms.tasks.OnCompleteListener
import com.google.android.gms.tasks.Task
import com.google.firebase.database.*

class MedicalReport() : AppCompatActivity() {
    var editTextTitle: EditText? = null
    var editTextDoctorName: EditText? = null
    var editTextResult: EditText? = null
    var editTextConclusion: EditText? = null
    private val myArrayList = ArrayList<String?>()
    var myReport: DatabaseReference? = null
    private var patientEmail: String? = null
    var Reports: DatabaseReference? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_medical_report)
        editTextTitle = findViewById<View>(R.id.editTextTextTitleMedical) as EditText
        editTextDoctorName = findViewById<View>(R.id.editTextTextDoctorNameMedical) as EditText
        editTextResult = findViewById<View>(R.id.editTextTextTestResultMedical) as EditText
        editTextConclusion = findViewById<View>(R.id.editTextTextConclusionMedical) as EditText
        Reports = FirebaseDatabase.getInstance().getReference("MedicalReport")
        myArrayList.add("Select One")
        myReport = FirebaseDatabase.getInstance().reference.child("Patients")
        myReport!!.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                for (snapshot: DataSnapshot in dataSnapshot.children) {
                    myArrayList.add(snapshot.child("emailAddress").value.toString())
                }
            }

            override fun onCancelled(error: DatabaseError) {}
        })
        val adapter1: ArrayAdapter<*> =
            ArrayAdapter<Any?>(this, android.R.layout.simple_spinner_item,
                myArrayList as List<Any?>
            )
        adapter1.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        val spinnerEmails = findViewById<View>(R.id.spinnerPatientMedical) as Spinner
        spinnerEmails.adapter = adapter1
        spinnerEmails.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(adapterView: AdapterView<*>?, view: View, i: Int, l: Long) {
                val myText: String = spinnerEmails.selectedItem.toString().trim { it <= ' ' }
                if (myText === "Select One".trim { it <= ' ' }) {
                    //Nothing
                } else {
                    Toast.makeText(
                        this@MedicalReport,
                        "You have Selected $myText",
                        Toast.LENGTH_SHORT
                    ).show()
                    patientEmail = myText
                }
            }

            override fun onNothingSelected(adapterView: AdapterView<*>?) {}
        }
        val myReport = findViewById<View>(R.id.buttonSendReportMedical) as Button
        myReport.setOnClickListener(View.OnClickListener {
            if (patientEmail!!.isEmpty()) {
                Toast.makeText(
                    this@MedicalReport,
                    "FIRST SELECT A PATIENT FROM ABOVE",
                    Toast.LENGTH_SHORT
                ).show()
            } else {
                setReport(patientEmail)
            }
        })
    }

    private fun setReport(myKey: String?) {
        var emailKey = ""
        val Counter = myKey!!.length
        for (a in 0 until Counter) {
            if (myKey[a] == '@') {
                break
            } else {
                emailKey = emailKey + myKey[a]
            }
        }
        emailKey = emailKey.trim { it <= ' ' }
        val title = editTextTitle!!.text.toString().trim { it <= ' ' }
        val doctorName = editTextDoctorName!!.text.toString().trim { it <= ' ' }
        val testResult = editTextResult!!.text.toString().trim { it <= ' ' }
        val conclusion = editTextConclusion!!.text.toString().trim { it <= ' ' }
        if (title.isEmpty()) {
            editTextTitle!!.error = "REPORT TITLE REQUIRED"
            editTextTitle!!.requestFocus()
        } else if (doctorName.isEmpty()) {
            editTextDoctorName!!.error = "DOCTOR NAME REQUIRED"
            editTextDoctorName!!.requestFocus()
        } else if (testResult.isEmpty()) {
            editTextResult!!.error = "TEST RESULT CANNOT BE BLANK"
            editTextResult!!.requestFocus()
        } else if (conclusion.isEmpty()) {
            editTextConclusion!!.error = "CONCLUSION IS REQUIRED"
            editTextConclusion!!.requestFocus()
        } else {
            val finalEmailKey = emailKey
            Reports!!.child(emailKey).get()
                .addOnCompleteListener(object : OnCompleteListener<DataSnapshot?> {
                    override fun onComplete(task: Task<DataSnapshot?>) {
                        if (task.isSuccessful) {
                            //String title, String doctorName, String testResult, String conclusion
                            val myReportsClass =
                                ReportClass(title, doctorName, testResult, conclusion)
                            Reports!!.child(finalEmailKey).setValue(myReportsClass)
                            //Toast.makeText(, "", Toast.LENGTH_SHORT).show();
                            Toast.makeText(
                                this@MedicalReport,
                                "REPORT SAVED SUCCESSFULLY",
                                Toast.LENGTH_SHORT
                            ).show()
                        }
                    }
                })
        }
    }
}